src/core/tuning.cpp

```diff
@@ -39,36 +37,63 @@
 #include <mpi.h>
 
 #include <algorithm>
-#include <cmath>
+#include <functional>
+#include <string>
+
+std::string TuningFailed::get_first_error() const {
+  using namespace ErrorHandling;
+  auto const queued_warnings = mpi_gather_runtime_errors_all(this_node == 0);
+  auto message = std::string("tuning failed: an exception was thrown while "
+                             "benchmarking the integration loop");
+  for (auto const &warning : queued_warnings) {
+    if (warning.level() == RuntimeError::ErrorLevel::ERROR) {
+      message += " (" + warning.what() + ")";
+      break;
+    }
+  }
+  return message;
+}
 
-double time_force_calc(int int_steps) {
+static void check_statistics(Utils::Statistics::RunningAverage<double> &acc) {
+  if (acc.avg() <= 5 * MPI_Wtick()) {
+    runtimeWarningMsg()
+        << "Clock resolution is too low to reliably time integration.";
+  }
+  if (acc.sig() >= 0.1 * acc.avg()) {
+    runtimeWarningMsg() << "Statistics of tuning samples is very bad.";
+  }
+}
+
+static void throw_on_error() {
+  auto const n_errors = check_runtime_errors_local();
+  if (boost::mpi::all_reduce(comm_cart, n_errors, std::plus<>()) != 0) {
+    throw TuningFailed{};
+  }
+}
+
+double benchmark_integration_step(int int_steps) {
   Utils::Statistics::RunningAverage<double> running_average;
 
-  if (mpi_integrate(0, 0))
-    return -1;
+  integrate(0, 0);
+  throw_on_error();
 
   /* perform force calculation test */
   for (int i = 0; i < int_steps; i++) {
-    const double tick = MPI_Wtime();
-
-    if (mpi_integrate(0, -1))
-      return -1;
-
-    const double tock = MPI_Wtime();
+    auto const tick = MPI_Wtime();
+    integrate(0, -1);
+    auto const tock = MPI_Wtime();
     running_average.add_sample((tock - tick));
+    throw_on_error();
   }
 
-  if (running_average.avg() <= 5 * MPI_Wtick()) {
-    runtimeWarningMsg()
-        << "Clock resolution is too low to reliably time integration.";
-  }
-
-  if (running_average.sig() >= 0.1 * running_average.avg()) {
-    runtimeWarningMsg() << "Statistics of tuning samples is very bad.";
+  if (this_node == 0) {
+    check_statistics(running_average);
   }
 
-  /* MPI returns s, return value should be in ms. */
-  return 1000. * running_average.avg();
+  /* MPI returns in seconds, returned value should be in ms. */
+  auto retval = 1000. * running_average.avg();
+  boost::mpi::broadcast(comm_cart, retval, 0);
+  return retval;
```

Comment:
Am i right that it wasn't necessary to broadcast this value before because the timing only ran on the head-node?

Reply:
that's correct; now all MPI ranks must have the same timings, otherwise the bisection will diverge